crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -259,17 +302,247 @@ where
         }
     }
 
+    pub fn transfer_object(
+        &mut self,
+        obj: ObjectValue,
+        addr: SuiAddress,
+    ) -> Result<(), ExecutionError> {
+        self.additional_transfers.push((addr, obj));
+        Ok(())
+    }
+
+    pub fn new_package(
+        &mut self,
+        modules: Vec<move_binary_format::CompiledModule>,
+    ) -> Result<(), ExecutionError> {
+        // wrap the modules in an object, write it to the store
+        let package_object = Object::new_package(
+            modules,
+            self.tx_context.digest(),
+            self.protocol_config.max_move_package_size(),
+        )?;
+        self.new_packages.push(package_object);
+        Ok(())
+    }
+
     pub fn push_command_results(&mut self, results: Vec<Value>) -> Result<(), ExecutionError> {
         assert_invariant!(
             self.borrowed.values().all(|is_mut| !is_mut),
             "all mut borrows should be restored"
         );
         // clear borrow state
         self.borrowed = HashMap::new();
-        self.results.push(results.into_iter().map(Some).collect());
+        self.results
+            .push(results.into_iter().map(ResultValue::new).collect());
         Ok(())
     }
 
+    /// Determine the object changes and collect all user events
+    pub fn finish(self) -> Result<ExecutionResults, ExecutionError> {
+        use sui_types::error::convert_vm_error;
+        let Self {
+            protocol_config,
+            vm,
+            state_view,
+            tx_context,
+            gas_status,
+            session,
+            additional_transfers,
+            new_packages,
+            gas,
+            mut inputs,
+            results,
+            user_events,
+            ..
+        } = self;
+        let tx_digest = tx_context.digest();
+        let sender = tx_context.sender();
+        let mut additional_writes = BTreeMap::new();
+        let mut input_object_metadata = BTreeMap::new();
+        // check for unused inputs
+        // track by-ref input objects as being written
```

Comment:
I see mut ref input objects as being treated as written, is that what this comment is referring to?

Reply:
I will rephrase

Reply:
Rephrasing also caught a bug :)